driver-scala/src/main/scala/org/mongodb/scala/internal/ZipObservable.scala

```diff
@@ -56,38 +60,61 @@ private[scala] case class ZipObservable[T, U](
         }
 
         override def onSubscribe(subscription: Subscription): Unit = {
-          if (firstSub) {
-            observable1Subscription = Some(subscription)
+          if (isLeftSub) {
+            leftSubscription = Some(subscription)
           } else {
-            observable2Subscription = Some(subscription)
+            rightSubscription = Some(subscription)
           }
 
-          if (observable1Subscription.nonEmpty && observable2Subscription.nonEmpty) {
+          if (leftSubscription.nonEmpty && rightSubscription.nonEmpty) {
             observer.onSubscribe(jointSubscription)
           }
         }
 
         override def onComplete(): Unit = {
-          if (!firstSub) {
-            terminated = true
-            observer.onComplete()
-          }
+          markCompleted(isLeftSub)
+          processNext(observer)
         }
 
         override def onNext(tResult: A): Unit = {
+          if (isLeftSub) leftCounter.incrementAndGet() else rightCounter.incrementAndGet()
           counter += 1
           queue.add((counter, tResult))
-          if (!firstSub) processNext(observer)
+          processNext(observer)
         }
       }
     }
 
-    private def processNext(observer: Observer[_ >: (T, U)]): Unit = {
-      (thisQueue.peek, thatQueue.peek) match {
-        case ((k1: Long, _), (k2: Long, _)) if k1 == k2 => observer.onNext((thisQueue.poll()._2, thatQueue.poll()._2))
+    private def markCompleted(isLeftSub: Boolean): Unit = synchronized {
+      if (isLeftSub) {
+        completedLeft = true
+      } else {
+        completedRight = true
+      }
+    }
+
+    private def completed(): Unit = synchronized {
+      if (!terminated) {
+        terminated = true
+        leftSubscription.foreach(_.unsubscribe())
```

Comment:
Can we assert anything about these unsubscribes happening?

Reply:
Not easily as the subscriptions aren't caught anywhere. It just allows any observables to potentially clean up resources as they may never actually complete.